### PR TITLE
Implement bullets

### LINKED ISSUE #55: Add bullets
Units should be able to fire bullets.

### Questions

 * Should they have a range?
 * Should they have a speed or hit just instant (laser-like)?
 * Should units autofire?
 * Should units calculate how to fire on moving targets?

### Tasks

- [x] Add bullets to `GameState`
- [ ] Add collision detection for bullets
- [x] Implement fire command
- [ ] Handle killed Units

### PR DESCRIPTION
**Pull request by rnestler**
Fixes #55

### CODE REVIEW

**Comment by dbrgn** [comment]
❇️ ✨ 🔫

File: src/client/mod.rs [hunk 1]
```diff
@@ -263,7 +265,9 @@ impl App {
         if let Some(game_state) = game_state_option {
             self.game_state = game_state;
         } else {
-            self.game_state.update(args.dt*1000.0);
+            if let Some(ref world) = self.world_state {
+                self.game_state.update(world, args.dt*1000.0);
```
Comment (dbrgn) [inline src/client/mod.rs:269]: Maybe we should always add spaces around operators?

File: src/client/mod.rs [hunk 2]
```diff
@@ -359,4 +366,11 @@ impl App {
             commands.push_back(Command::Move(*u, position));
         }
     }
+
+    pub fn shoot_selected(&mut self) {
```
Comment (dbrgn) [inline src/client/mod.rs]: Maybe rename to `shoot`?

File: src/server/mod.rs [hunk 3]
```diff
@@ -118,14 +118,13 @@ pub fn handle_client(mut stream: TcpStream,
 
                     // Send ServerHello message
                     let encoded: Vec<u8> = serialize(
-                        &Message::ServerHello(player_id, world.lock().unwrap().clone()),
+                        &Message::ServerHello(player_id, (*world).clone()),
```
Comment (dbrgn) [inline src/server/mod.rs]: Style: I'd use `.deref()` to avoid the parentheses.

File: src/server/mod.rs [hunk 4]
```diff
@@ -135,7 +134,7 @@ pub fn handle_client(mut stream: TcpStream,
 
                             // Send ServerHello message
                             let encoded: Vec<u8> = serialize(
-                                &Message::ServerHello(id, world_lock.clone()),
+                                &Message::ServerHello(id, (*world).clone()),
```
Comment (dbrgn) [inline src/server/mod.rs]: Here again. And probably in some other places too :)

File: src/state.rs [hunk 5]
```diff
@@ -123,10 +163,53 @@ impl GameState {
...
+            // iterating
+            let mut units = mem::replace(&mut player.units, vec![]);
+            for unit in units.iter_mut() {
+                if unit.health > 0 {
+                    unit.update(dt);
+                    player.units.push(unit.clone());
+                } else {
+                    println!("killed unit!");
+                }
+            }
```
Comment (dbrgn) [inline src/state.rs]: This could be done cleanly with an iterator and filter. Maybe also for the bullets?

**Review by dbrgn — APPROVED** [review]
In general it's great, but maybe you can fix these issues.

File: src/state.rs (hunk 5 above)
Comment (rnestler) [reply src/state.rs]: Done.

File: src/server/mod.rs (hunk 4 above)
Comment (rnestler) [reply src/server/mod.rs]: Done.

File: src/server/mod.rs (hunk 3 above)
Comment (rnestler) [reply src/server/mod.rs]: Done.

File: src/client/mod.rs (hunk 2 above)
Comment (rnestler) [reply src/client/mod.rs]: Done.

**Comment by dbrgn** [comment]
:sparkles:

File: src/client/mod.rs (hunk 1 above)
Comment (rnestler) [reply src/client/mod.rs:269]: I think I'll run a complete rust-fmt run in a later PR and enable CI for it. But lets first merge the existing PRs.

**Pull request merged**